From c9bed48f65b83a2387a1606325e326f400ed8af2 Mon Sep 17 00:00:00 2001
From: MaxHearnden <maxoscarhearnden@gmail.com>
Date: Sat, 7 Jun 2025 05:59:32 +0100
Subject: [PATCH] Match address when searching for systemd sockets

---
 services/listen_dnsport.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/services/listen_dnsport.c b/services/listen_dnsport.c
index a3b4d4941..3d9dc649a 100644
--- a/services/listen_dnsport.c
+++ b/services/listen_dnsport.c
@@ -219,7 +219,7 @@ systemd_get_activated(int family, int socktype, int listen,
 	}
 
 	for(i = 0; i < r; i++) {
-		if(sd_is_socket(SD_LISTEN_FDS_START + i, family, socktype, listen)) {
+		if(sd_is_socket_sockaddr(SD_LISTEN_FDS_START + i, socktype, addr, addrlen, listen)) {
 			s = SD_LISTEN_FDS_START + i;
 			break;
 		}
-- 
2.49.0

